build: Add CMake-based build system (1 of N)

hebasto opened the pull request:
This is split from https://github.com/bitcoin/bitcoin/pull/25797 which is a full implementation with an Autotools feature parity and has more than 80 commits.

A bit [rehashed](https://github.com/bitcoin/bitcoin/pull/25797#issue-1330985812) list of benefits of a CMake-based build system over an Autotools-based one:

## General benefits

1. [Integration](https://www.kitware.com/static-checks-with-cmake-cdash-iwyu-clang-tidy-lwyu-cpplint-and-cppcheck/) with developer tools like clang-tidy, IWYU etc.

2. CTest tool with a bunch of useful features, including coverage analysis, integration with sanitizers and memory check tools like valgrind.

3. [CPack](https://cmake.org/cmake/help/book/mastering-cmake/chapter/Packaging%20With%20CPack.html), a powerful packaging tool.

4. Integration with IDEs.

5. The concept of a "target with assosiated properties" is better than Autotools' global variables in many ways (robustness, readability, maintanability).

6. _[social]_ A great community. CMake is actively developed.


## The Bitcoin Core specific benefits

A preliminary note. Two arguments (a) "nobody cares about X platform" and (b) "nobody cares about Y binary" are out of this PR scope, as long as the X platform is promised to be supported by the Bitcoin Core project, and the Y binary is a part of the Bitcoin Core project.

7. Native Windows support. No longer need to maintain `build_msvc` subdirectory.

8. Correctness with Windows DLLs out-of-the-box. In comparison, the current build system has 3 ([three!](https://github.com/bitcoin/bitcoin/issues/25008#issuecomment-1113964111)) hacks of libtool. And it is still broken:
- https://github.com/bitcoin/bitcoin/issues/19772
- https://github.com/bitcoin/bitcoin/issues/25008

9. CMake is a native build system of [Qt 6](https://doc.qt.io/qt-6/cmake-manual.html).

10. _[social]_ It is [expected](https://github.com/bitcoin/bitcoin/pull/25797#issuecomment-1408772481) that more current and new developers are/will be able/willing to review/contribute/maintain CMake code rather Autotools stuff.

## Roadmap

There are two goals:
- merge all CMake code by v25.0 and test it
- switch to the CMake-based build system by v26.0

## Implementation details

In particular, this PR establishes the minimum required CMake [version](https://gitlab.kitware.com/cmake/community/-/wikis/CMake-Versions-on-Linux-Distros). As it looks non-productive to support Ubuntu Bionic as a build platform, version [3.13](https://cmake.org/cmake/help/latest/release/3.13.html) has been chosen. Benefits of v3.13 over v3.10 which affect our CMake code directly are as [follow](https://github.com/hebasto/bitcoin/blob/220807-cmake/CMakeLists.txt):
```CMake
#  - 3.11: add_library() and add_executable() commands can now be called without any sources
#          and will not complain as long as sources are added later via the target_sources()
#          command.
#  - 3.11: The target_*() commands learned to set the INTERFACE properties on Imported Targets.
#  - 3.12: The add_compile_definitions() command was added to set preprocessor definitions
#          at directory level. This supersedes add_definitions().
#  - 3.12: If the CONFIGURE_DEPENDS flag is specified, CMake will add logic to the main
#          build system check target to rerun the flagged GLOB commands at build time.
#  - 3.12: The target_link_libraries() command now supports Object Libraries.
#  - 3.12: A new $<TARGET_EXISTS:...> generator expression has been added.
#  - 3.12: A new FindPython3 module has been added to provide a new way to locate python
#          environments.
#  - 3.13: The install(TARGETS) command learned to install targets created outside the current
#          directory.
#  - 3.13: The target_link_options() command was created to specify link options for targets
#          and their dependents.
#  - 3.13: The target_link_libraries() command may now be called to modify targets created
#          outside the current directory.
```

## Notes for reviewers

To configure a build system, one can run in their local repo root:
- on Linux / *BSD / macOS:
```sh
cmake -S . -B  build
```
- on Windows:
```cmd
cmake -G "Visual Studio 17 2022" -A x64 -S . -B build
```

To re-configure, it is recommended to use the clean build tree, for example:
```sh
rm -rf build
```

For now, the only artifact we explicitly create in the build tree is a `bitcoin-config.h` header:
```sh
cat build/src/config/bitcoin-config.h
```

hebasto commented:
Friendly ping people who commented on #25797:

@achow101 
@adam2k
@dongcarl 
@fanquake
@jarolrod 
@luke-jr 
@MarcoFalke
@ryanofsky
@sipsorcery 
@TheCharlatan
@theStack
@theuni
@vasild
@willcl-ark

vasild commented on CMakeLists.txt:
```diff
@@ -0,0 +1,71 @@
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+cmake_minimum_required(VERSION 3.13)
+
+if(CMAKE_VERSION VERSION_GREATER_EQUAL 3.15)
+  # MSVC runtime library flags are selected by the CMAKE_MSVC_RUNTIME_LIBRARY abstraction.
+  cmake_policy(SET CMP0091 NEW)
+  # MSVC warning flags are not in CMAKE_<LANG>_FLAGS by default.
+  cmake_policy(SET CMP0092 NEW)
+endif()
+if(CMAKE_VERSION VERSION_GREATER_EQUAL 3.20)
+  # MSVC RTTI flag /GR is not added to CMAKE_CXX_FLAGS by default.
+  cmake_policy(SET CMP0117 NEW)
+endif()
```
nit: CMake 3.13 is from Nov 2018. Is there any reason to support that? Why not just `cmake_minimum_required(VERSION 3.20)` and remove those `if`s? (I guess some laggish linux distro still ships with 3.13?)

hebasto replied:
> I guess some laggish linux distro still ships with 3.13?

Debian 10 Buster.

hebasto replied:
> nit: CMake 3.13 is from Nov 2018.

The [draft](https://github.com/bitcoin/bitcoin/pull/25797) supports even CMake 3.10 which is shipped with Ubuntu Bionic :)

hebasto replied:
Related: today's IRC meeting [log](https://www.erisian.com.au/bitcoin-core-dev/log-2023-02-09.html#l-270).

vasild commented on CMakeLists.txt:
```diff
@@ -21,7 +21,11 @@ project("Bitcoin Core"
-if(MSVC)
+# Configurable options.
+# When adding a new option, end the <help_text> with a full stop for consistency.
+option(CXX20 "Enable compilation in C++20 mode." OFF)
+
+if(CXX20 OR MSVC)
```
nit: this would enable C++20 unconditionally on MSVC. Maybe the option should not exist on MSVC? For Windows users, it would be strange to have an option which does nothing and when CXX20=OFF, then we still compile in C++20 mode.

C++17 seems to be supported on MSCV in `configure.ac`.

hebasto replied:
> C++17 seems to be supported on MSCV in `configure.ac`.

MSVC build configuration is defined in the `build_msvc` directory, not by `configure.ac`.

And C++20 for MSVC was introduced in 88ec5d40dcf5d9f95217b123b48203b2f334c0a1 from #25472.

hebasto replied:
> Maybe the option should not exist on MSVC? 

Thanks! Going to fix this.

hebasto replied:
> > Maybe the option should not exist on MSVC?
> 
> Thanks! Going to fix this.

[Updated](https://github.com/bitcoin/bitcoin/pull/27060#issuecomment-1424023133).

vasild commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+configure_file(${CMAKE_SOURCE_DIR}/cmake/bitcoin-config.h.in config/bitcoin-config.h @ONLY)
```
This would generate the config file in `config/bitcoin-config.h` whereas with autotools it would be in `src/config/bitcoin-config.h`. I guess this is ok (as long as the proper include flags are passed)?

hebasto replied:
It generates `${CMAKE_CURRENT_BINARY_DIR}/config/bitcoin-config.h`, which is equivalent to `${CMAKE_BINARY_DIR}/src//config/bitcoin-config.h`.

The related `include_directories()` goes two lines below :)

vasild reviewed (approved):
ACK 044ff46008645ee7b77575d7d68ac12dc6b99d14

The first two commits may as well be squashed into one. Some non-blocker questions below.

hebasto commented:
Updated 044ff46008645ee7b77575d7d68ac12dc6b99d14 -> 1d4cec0f80d01a06418b59b1563da72996a7bd79 ([pr27060.02](https://github.com/hebasto/bitcoin/commits/pr27060.02) -> [pr27060.03](https://github.com/hebasto/bitcoin/commits/pr27060.03), [diff](https://github.com/hebasto/bitcoin/compare/pr27060.02..pr27060.03)):

- addressed @vasild's [comment](https://github.com/bitcoin/bitcoin/pull/27060#discussion_r1101125181)

> The first two commits may as well be squashed into one.

Squashed.

vasild reviewed (approved):
ACK 1d4cec0f80d01a06418b59b1563da72996a7bd79

sedited reviewed (approved):
ACK 1d4cec0f80d01a06418b59b1563da72996a7bd79

And to be clear, also approach ACK for doing this piecemeal and the targeted timeline.

theStack reviewed (approved):
ACK 1d4cec0f80d01a06418b59b1563da72996a7bd79

Tested with OpenBSD 7.2 and cmake 3.24.2, verified that the generated `src/config/bitcoin-config.h` file in the build directory is a subset of the one generate by autotools. Can't relate to the MSVC-related changes, since I don't have a Windows build environment available.

jarolrod reviewed (approved):
ACK 1d4cec0f80d01a06418b59b1563da72996a7bd79

tested on macOS 12.6 M1, cannot currently test for windows.

MarcoFalke commented on cmake/bitcoin-config.h.in:
```diff
@@ -0,0 +1,44 @@
+#define CLIENT_VERSION_MINOR @CLIENT_VERSION_MINOR@
+
+/* Version build. */
+#define CLIENT_VERSION_BUILD @CLIENT_VERSION_BUILD@
```
Any reason to change the order and at the same the time the docstrings? This makes it harder to review the diff:


```diff
diff --git a/tmp/a b/tmp/b
index 13d6b32ec..bd846a352 100644
--- a/tmp/a
+++ b/tmp/b
@@ -1,442 +1,44 @@
-/* src/config/bitcoin-config.h.  Generated from bitcoin-config.h.in by configure.  */
-/* src/config/bitcoin-config.h.in.  Generated from configure.ac by autoheader.  */
+// Copyright (c) 2023 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
 #ifndef BITCOIN_CONFIG_H
-
 #define BITCOIN_CONFIG_H
 
-/* Define if building universal (internal helper macro) */
-/* #undef AC_APPLE_UNIVERSAL_BUILD */
+/* Define to the full name of this package. */
+#define PACKAGE_NAME "Bitcoin Core"
 
-/* Define this symbol if type char equals int8_t */
-/* #undef CHAR_EQUALS_INT8 */
+/* Define to the version of this package. */
+#define PACKAGE_VERSION "24.99.0"
 
-/* Version Build */
-#define CLIENT_VERSION_BUILD 0
-
-/* Version is release */
-#define CLIENT_VERSION_IS_RELEASE false
-
-/* Major version */
+/* Major version. */
 #define CLIENT_VERSION_MAJOR 24
 
-/* Minor version */
+/* Minor version. */
 #define CLIENT_VERSION_MINOR 99
 
-/* Copyright holder(s) before %s replacement */
-#define COPYRIGHT_HOLDERS "The %s developers"
+/* Version build. */
+#define CLIENT_VERSION_BUILD 0
 
-/* Copyright holder(s) */
-#define COPYRIGHT_HOLDERS_FINAL "The Bitcoin Core developers"
-
-/* Replacement for %s in copyright holders string */
-#define COPYRIGHT_HOLDERS_SUBSTITUTION "Bitcoin Core"
-
-/* Copyright year */
-#define COPYRIGHT_YEAR 2023
-
-/* Define this symbol to build code that uses ARMv8 SHA-NI intrinsics */
-/* #undef ENABLE_ARM_SHANI */
-
...

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/27060#issuecomment-1425893451).

MarcoFalke commented on CMakeLists.txt:
```diff
@@ -0,0 +1,91 @@
+cmake_dependent_option(CXX20 "Enable compilation in C++20 mode." OFF "NOT MSVC" ON)
+
+if(CXX20)
+  set(CMAKE_CXX_STANDARD 20)
```
I changed this to 26 for testing, but didn't run into an issue on cmake 3.13. Is this expected?

hebasto replied:
> I changed this to 26 for testing, but didn't run into an issue on cmake 3.13. Is this expected?

Mind providing the exact command?

MarcoFalke replied:
```
# grep CMAKE_CXX_STANDARD CMakeLists.txt && cmake --version && cmake   -S . -B ./build7  
  set(CMAKE_CXX_STANDARD 26)
  set(CMAKE_CXX_STANDARD 26)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
cmake version 3.13.4

CMake suite maintained and supported by Kitware (kitware.com/cmake).
-- The CXX compiler identification is GNU 8.3.0
-- The C compiler identification is GNU 8.3.0
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/cc
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done


Configure summary
=================
Preprocessor defined macros ........... 
C compiler ............................ /usr/bin/cc
CFLAGS ................................ 
C++ compiler .......................... /usr/bin/c++
CXXFLAGS .............................. 
Common compile options ................ 
Common link options ................... 
Build type:
 - CMAKE_BUILD_TYPE ................... 
 - CFLAGS ............................. 
 - CXXFLAGS ........................... 
 - LDFLAGS for executables ............ 
 - LDFLAGS for shared libraries ....... 


-- Configuring done
-- Generating done
-- Build files have been written to: /bitcoin-core/build7

hebasto replied:
The [`CMAKE_CXX_STANDARD`](https://cmake.org/cmake/help/latest/variable/CMAKE_CXX_STANDARD.html) variable just holds a default value for [`CXX_STANDARD`](https://cmake.org/cmake/help/latest/prop_tgt/CXX_STANDARD.html) target property, which CMake will actually check.

To see such behavior, please consider applying a diff as follows:
```diff
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -89,3 +89,5 @@ else()
   message(" - LDFLAGS for shared libraries ....... ${CMAKE_SHARED_LINKER_FLAGS_RELEASE}")
 endif()
 message("\n")
+
+add_library(test_lib OBJECT src/addrdb.cpp)
```

hebasto commented:
Updated 1d4cec0f80d01a06418b59b1563da72996a7bd79 -> f9e1e7264f925690db74abd3efe7f1a3e10f2542 ([pr27060.03](https://github.com/hebasto/bitcoin/commits/pr27060.03) -> [pr27060.04](https://github.com/hebasto/bitcoin/commits/pr27060.04), [diff](https://github.com/hebasto/bitcoin/compare/pr27060.03..pr27060.04)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/27060#discussion_r1102774935)

willcl-ark commented:
I notice in the generated file _build/CMakeCache.txt_ that I have the following:

```
//Choose the type of build, options are: None Debug Release RelWithDebInfo
// MinSizeRel ...
CMAKE_BUILD_TYPE:STRING=
```

Should this be set to `Release` by default so that we pick up Release optimisations?

hebasto commented:
> I notice in the generated file _build/CMakeCache.txt_ that I have the following:
> 
> ```
> //Choose the type of build, options are: None Debug Release RelWithDebInfo
> // MinSizeRel ...
> CMAKE_BUILD_TYPE:STRING=
> ```
> 
> Should this be set to `Release` by default so that we pick up Release optimisations?

You are correct! Those commits are coming :)
See 290f1e36ddf697c0825aa3340ff50ebc83a3b572 and 9d8bf0b1302651e825081ce8186822b510b3b83c from #25797.

willcl-ark commented:
> You are correct! Those commits are coming :) See [290f1e3](https://github.com/bitcoin/bitcoin/commit/290f1e36ddf697c0825aa3340ff50ebc83a3b572) and [9d8bf0b](https://github.com/bitcoin/bitcoin/commit/9d8bf0b1302651e825081ce8186822b510b3b83c) from #25797.

Ah I see. In that case LGTM!

willcl-ark reviewed (approved):
ACK f9e1e7264

fanquake commented:
> merge all CMake code by v25.0

NACK merging this prior to the 25.x branch off. I don't understand why it's being split up, and we're now trying to merge changes that aren't even usable to compile Core? (also resulting in the confusion above). How many steps in this going to be broken into? What is the reason to get this merged now.

> and test it

Developers should at-least be testing, and becoming comfortable with CMake before this is merged, not after.

> switch to the CMake-based build system by v26.0

What does this mean? When are we removing Autotools? We aren't going to support two build systems in this repository at the same time. That is just going to draw out and overcomplicate what is already going to be a long and drawn-out (and breaking) process of supporting both (via release branches).

hebasto commented:
> I don't understand why it's being split up

It is split up because #25797 diff seems unreasonable large for reviewing.

> How many steps in this going to be broken into?

I'd say not "how many steps" but "how large diff for every step". To be done, things must be doable.

> > switch to the CMake-based build system by v26.0
> 
> What does this mean?

Use CMake-based build system for a release.

> When are we removing Autotools?

Any time developers are comfortable with.

> We aren't going to support two build systems in this repository at the same time.

Good.

theuni commented:
> > merge all CMake code by v25.0
> 
> NACK merging this prior to the 25.x branch off. I don't understand why it's being split up, and we're now trying to merge changes that aren't even usable to compile Core? (also resulting in the confusion above). How many steps in this going to be broken into? What is the reason to get this merged now.

+1 this take. NACK for incomplete/parallel buildsystems in the v25 branch. I'd really like to review this PR (and the others) thoroughly and thoughtfully now that you consider them merge-ready, but not under threat of "whatever parts of this that are ready are being merged during this release cycle, better fix up the chunks as best we can first". Without everyone dogfooding at once, I'm afraid we might not catch high-level problems before moving on to the next chunk in the series to review.

Merging for v25 would mean maintaining parallel changes for that cycle. I think that's just a recipe for disaster. IMO the most straightforward way forward would be to merge at the beginning of a release cycle (v26 if it's ready by then) and deprecate autotools at the same time, potentially even removing it before tag. That way everyone switches at once and new features don't have to be shoved into the old buildsystem, it would simply be frozen in time.

So... what's the rush? I realize that this is a monster to maintain, but it's just the nature of such a big change.

Speaking more helpfully and practically, this change may be big enough to warrant an out-of-the-ordinary workflow: maybe either working on a CMake branch in the Core repo, or hooking CI up to an additional repo. Since (I believe) the desire to merge for v25 comes from the desire for more eyes and testing, maybe providing a more "official" way of trying out CMake pre-merge would be enough?

vasild reviewed (approved):
ACK f9e1e7264f925690db74abd3efe7f1a3e10f2542

The parent PR is too huge to review. If that is to be merged at once, then this is literally asking to merge it without proper review. Thus, it makes sense to me to split it. Yes, this means in the intermediate steps it is not going to be possible to build using CMake. I think that is ok.

If that is going to help, maybe put everything cmake related in a subdirectory like e.g. `./cmake_incomplete/CMakeLists.txt` and/or requiring that it is run like `cmake -D I_UNDERSTAND_THIS_IS_INCOMPLETE_UNMAINTAINED_AND_NOT_OFFICIALLY_SUPPORTED_CMAKE_BUILD_FOR_DEVS_ONLY=1` and aborting with an appropriate message otherwise?

> We aren't going to support two build systems in this repository at the same time.

I agree that better be avoided. Merging this and subsequent PRs does not mean that a CMake build system is going to be supported - `doc/build*` still mentions autotools solely and nothing about CMake.

hebasto commented:
@theuni 

Thank you for your helpful comment.

If I understand you correctly, you are suggesting to follow the same path as #2943 did, right?

If so, should that huge unstructured patch touch (a) Guix-related stuff, (b) CI-related stuff? Or Guix and CI changes should go separately?

> but not under threat

Sorry, but I did not mean any threats.

TheCharlatan commented:
ACK f9e1e7264f925690db74abd3efe7f1a3e10f2542

I was about to post what @vasild just suggested and second his approach.

theuni commented:
@hebasto Yes. I'm going to review/test the CMake work (starting with this PR) in earnest this week. Until then, I'll hold off on further commenting/judgement.

I do stand by my comment on timing though. Getting this merged "as-is" for v25 is a bad idea imo. I suggest that we create a staging branch/repo for reviewing and merging chunks at a time, with the goal of merging from staging to master after v26 branch. Similar to the workflow (though not necessarily the timing) @sipa used for segwit.

hebasto commented on CMakeLists.txt:
```diff
@@ -0,0 +1,76 @@
+# Configurable options.
+# When adding a new option, end the <help_text> with a full stop for consistency.
+include(CMakeDependentOption)
+cmake_dependent_option(CXX20 "Enable compilation in C++20 mode." OFF "NOT MSVC" ON)
```
> Why is c++20 turned on automatically for MSVC? Afaics autotools doesn't do that.

https://github.com/bitcoin/bitcoin/pull/27060#discussion_r1101149891:

MSVC build configuration is defined in the `build_msvc` directory, not in `configure.ac`.

And C++20 for MSVC was introduced in 88ec5d40dcf5d9f95217b123b48203b2f334c0a1 from #25472.

theuni replied:
Thanks. I don't understand that change, but I'll take my complaints to that PR :)

hebasto replied:
It originates from #24531.

hebasto commented on CMakeLists.txt:
```diff
@@ -0,0 +1,76 @@
+
+cmake_minimum_required(VERSION 3.13)
+
+if(CMAKE_VERSION VERSION_GREATER_EQUAL 3.15)
```
https://cmake.org/cmake/help/latest/manual/cmake-policies.7.html:

> Policies in CMake are used to preserve backward compatible behavior across multiple releases.

theuni replied:
Understood for policies. But that's not to mean that it's not possible to misuse or misunderstand, which is why I'm suggesting the note. For example, one could imagine something naive/broken like:
```CMake
if(CMAKE_VERSION VERSION_GREATER_EQUAL 3.20)
  set(CMAKE_POSITION_INDEPENDENT_CODE ON)
endif()
```
which would provide better binaries for builders with newer CMake.

Maybe the note could say (if it's true?) that only `cmake_policy` should be set as a result of `CMAKE_VERSION` testing?

hebasto replied:
> But IMO we need a note to the effect of: "Changes which affect binary results may _not_ be quietly gated by CMake version".

Ah, you mean a comment in CMAke's code, right?

hebasto replied:
> > But IMO we need a note to the effect of: "Changes which affect binary results may _not_ be quietly gated by CMake version".
> 
> Ah, you mean a comment in CMake's code, right?

Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/27060#issuecomment-1429846434).

hebasto commented on CMakeLists.txt:
```diff
@@ -0,0 +1,91 @@
+project("Bitcoin Core"
+  VERSION 24.99.0
+  DESCRIPTION "Bitcoin client software"
+  LANGUAGES CXX C ASM
```
> ```
> export PATH=/opt/clang+llvm-14.0.0-x86_64-linux-gnu-ubuntu-18.04/bin:$PATH
> cmake -S . -B ./build2
> ```
> 
> Rather than picking up `/opt/clang+llvm-14.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang`, it finds `/usr/bin/cc`.

```
cmake -S . -B ./build2 \
  -DCMAKE_C_COMPILER=/opt/clang+llvm-14.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang \
  -DCMAKE_CXX_COMPILER=/opt/clang+llvm-14.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang++
```

theuni replied:
Of course. I was simply pointing out that in this case, `./configure && make` wouldn't map to `cmake -S . -B ./build && make -C build` as one might expect. So it's a question of expectations of current build parity vs CMake-ness.

vasild replied:
Do autotools really pick `/foo/bin/clang` (assuming `/foo/bin` is in `$PATH`)? I tried that but for me it picked `/usr/bin/c++` nevertheless, even though I had a `clang` executable in a directory earlier than `/usr/bin` in `$PATH`.

hebasto commented:
Updated f9e1e7264f925690db74abd3efe7f1a3e10f2542 -> 75c4333e484ffea433f32b48ca7d4a1dc819bb36 ([pr27060.04](https://github.com/hebasto/bitcoin/commits/pr27060.04) -> [pr27060.05](https://github.com/hebasto/bitcoin/commits/pr27060.05), [diff](https://github.com/hebasto/bitcoin/compare/pr27060.04..pr27060.05)):

- addressed @theuni's [comment](https://github.com/bitcoin/bitcoin/pull/27060#discussion_r1104821340)

hebasto commented on CMakeLists.txt:
```diff
@@ -0,0 +1,91 @@
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+set(CMAKE_CXX_EXTENSIONS OFF)
+
+set(CMAKE_POSITION_INDEPENDENT_CODE ON)
```
Thanks! [Fixed](https://github.com/bitcoin/bitcoin/pull/27060#issuecomment-1428512812).

hebasto replied:
[UPD](https://github.com/bitcoin/bitcoin/pull/27060#issuecomment-1429846434): Now it is covered by the version range in the `cmake_minimum_required()` command.

vasild commented on CMakeLists.txt:
```diff
@@ -0,0 +1,110 @@
+if(configure_warnings)
+    message("  ******\n")
+    foreach(warning IN LISTS configure_warnings)
+      message("  WARNING: ${warning}\n")
```
nit:

```suggestion
      message(WARNING "${warning}")
```

https://cmake.org/cmake/help/latest/command/message.html

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/27060#issuecomment-1429846434).

vasild reviewed (approved):
ACK 75c4333e484ffea433f32b48ca7d4a1dc819bb36

hebasto committed:
cmake: Add root `CMakeLists.txt` file

hebasto committed:
cmake: Add `config/bitcoin-config.h` support

hebasto commented:
Updated 75c4333e484ffea433f32b48ca7d4a1dc819bb36 -> 12758d3f69c4900f0d23df06e1ad746417428f3a ([pr27060.05](https://github.com/hebasto/bitcoin/commits/pr27060.05) -> [pr27060.06](https://github.com/hebasto/bitcoin/commits/pr27060.06)):

- rebased on top of the recent changes in the build system (#27016)
- `cmake_policy()` commands replaced with version range in `project()` one
- documented choice of versions
- add a note [suggested](https://github.com/bitcoin/bitcoin/pull/27060#discussion_r1104782068) by @theuni
- addressed @vasild's [comment](https://github.com/bitcoin/bitcoin/pull/27060#discussion_r1105862197)
- second commit diff minimized

adam2k commented:
This is great!  ACK: https://github.com/bitcoin/bitcoin/commit/12758d3f69c4900f0d23df06e1ad746417428f3a

I see from the `NACK` comments that the reviewers seem to mostly be pushing back on this portion:

> ## Roadmap
> There are two goals:
> 
> * merge all CMake code by v25.0 and test it
> * switch to the CMake-based build system by v26.0

In the spirit of having a complete git working history in the `master` branch, I understand the viewpoint.

Are there any other pieces (without overly complicating things) that can also be pulled in from into this PR to make this more "complete"?  If it's not possible then I agree it's not ideal, but it will make future review easier and to the point of other `ACK` reviewers this PR does not add anything about `cmake` within the docs.

hebasto commented:
@fanquake 

> 3\. [Comments like](https://github.com/bitcoin/bitcoin/pull/27060#issuecomment-1425962907):
>         > When are we removing Autotools?
>         
>         
>         > > Any time developers are comfortable with.
> 
> 
> are pretty hand wavy, and do not establish any sort of plan.

My comment is "pretty hand wavy" for the only reason -- I have no experience of introducing a new build system for a project like Bitcoin Core (#2943 happened before I ran my first Bitcoin Core node :smile:). That's why I'm asking you to share your vision of the optimal way to do it, which will avoid all drawbacks you've mentioned multiple times already.

Anyway, a discussion "how to introduce a new build system into the project" can be separated from another one "how a new build system is implemented".

---

> 1\. The parent PR (#25797) suffers from a lack of high-level review.

Well, the "100% feature parity" requirement was actively discussed among developers (@dongcarl, @theuni, @ryanofsky) there:
- https://github.com/bitcoin/bitcoin/pull/25797#issuecomment-1216939762
- https://github.com/bitcoin/bitcoin/pull/25797#issuecomment-1218365904
- https://github.com/bitcoin/bitcoin/pull/25797#issuecomment-1219758967

And, I think (probably mistakenly), it is implicitly expected in @MarcoFalke's https://github.com/bitcoin/bitcoin/pull/27060#discussion_r1102774935. I mean, when system introspection code will be added, there must be no diff between an Autotools-generated `bitcoin-config.h` and CMake-generated one.

---

> 2\. There is no urgency to do this.

True. No code must be merged before it has been reviewed and tested properly.

---

Will we be OK with this PR as a "reviewing stage"? I mean, adding more commits after reviewing and ACKing of previous ones, with optional squashing of already reviewed commits, and rebasing on top of the recent changes in the current build system in the master branch.

Doing in such a way we can get a reviewed CMake-based build system implementation ready to be merged at some moment in the future.

theuni commented:
> Will we be OK with this PR as a "reviewing stage"? I mean, adding more commits after reviewing and ACKing of previous ones, with optional squashing of already reviewed commits, and rebasing on top of the recent changes in the current build system in the master branch.
> 
> Doing in such a way we can get a reviewed CMake-based build system implementation ready to be merged at some moment in the future.

My comment above:
> I suggest that we create a staging branch/repo for reviewing and merging chunks at a time, with the goal of merging from staging to master after v26 branch. Similar to the workflow (though not necessarily the timing) @sipa used for segwit.

Again, I suggest a staging repo for review. It could just be your personal repo where you PR to a branch. Review happens in chunks (as you're doing now) but to that branch, that way we don't have the temporary clutter in the master repo. Then, once all the chunks have been individually merged and reviewed, it is submitted as one giant pr to master where it can be tested in full before final merge. This is essentially the [Dictator and Lieutenants Workflow](https://git-scm.com/book/en/v2/Distributed-Git-Distributed-Workflows) This would allow us to collaborate on PR chunks and keep track of issues, all without disrupting the main repo, or attracting cheerleading.

hebasto commented:
@theuni 

Thank you!

> Again, I suggest a staging repo for review. It could just be your personal repo where you PR to a branch. Review happens in chunks (as you're doing now) but to that branch, that way we don't have the temporary clutter in the master repo. Then, once all the chunks have been individually merged and reviewed, it is submitted as one giant pr to master where it can be tested in full before final merge. This is essentially the [Dictator and Lieutenants Workflow](https://git-scm.com/book/en/v2/Distributed-Git-Distributed-Workflows) This would allow us to collaborate on PR chunks and keep track of issues, all without disrupting the main repo, or attracting cheerleading.

Like [that](https://github.com/hebasto/bitcoin/pull/5)?

hebasto commented:
> So, right, I'd like to request that everyone who has been involved in reviewing (technically) the other CMake PRs to come along with me to reviewing over at @hebasto's staging repo. Since CMake is apparently so urgent, I'm sure we'll have no trouble recruiting review over there :p

Just in case, I've made an announcement in #bitcoin-core-builds IRC channel.

fanquake commented:
I think we can close this now that review is happening elsewhere, and we've still got the parent PR.
